sound/soc/intel/boards/sof_sdw_cs42l43.c

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Based on sof_sdw_rt5682.c
+// Copyright (c) 2023 Intel Corporation
+
+/*
+ *  sof_sdw_cs42l43 - Helpers to handle CS42L43 from generic machine driver
+ */
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <linux/input.h>
+#include <sound/jack.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#include <sound/cs42l43.h>
+#include <sound/control.h>
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
+#include <sound/soc-dapm.h>
+#include "sof_sdw_common.h"
+
+static const struct snd_soc_dapm_widget sof_widgets[] = {
+	SND_SOC_DAPM_SPK("Speaker", NULL),
+};
+
+static const struct snd_soc_dapm_route sof_map[] = {
+	{ "Speaker", NULL, "cs42l43 AMP1_OUT_P" },
+};
+
+static int cs42l43_rtd_init(struct snd_soc_pcm_runtime *rtd)
+{
+	struct snd_soc_component *component = asoc_rtd_to_codec(rtd, 0)->component;
+	struct mc_private *ctx = snd_soc_card_get_drvdata(rtd->card);
+	struct snd_soc_jack *jack = &ctx->sdw_headset;
+	struct snd_soc_card *card = rtd->card;
+	int ret;
+
+	dev_dbg(card->dev, "%s()\n", __func__);
+
+	card->components = devm_kasprintf(card->dev, GFP_KERNEL, "%s hs:cs42l43", card->components);
```

Comment:
this could be fun if we have cohen+jameson connected, we probably need to work on the 'component' string a bit more so that UCM files figure out how the headset codec and amplifiers are setup.

Reply:
> this could be fun if we have cohen+jameson connected, we probably need to work on the 'component' string a bit more so that UCM files figure out how the headset codec and amplifiers are setup.

I add a separated init callback for DMIC and add `"%s mic:cs42l43-dmic` to card->components.